File: cv_pkg/src/cv_image_node.py
#!/usr/bin/env python3
#conding: utf-8
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image
from cv_bridge import CvBridge
import cv2

bridge = None
node = None

def cam_rgb_callback(msg):
    global bridge, node

    try:
        cv_image = bridge.imgmsg_to_cv2(msg, desired_encoding="bgr8")
    except Exception as e:
        node.get_logger().error(f"cv_bridge exception: {e}")
        return

    # Display the image using OpenCV
    cv2.imshow("RGB", cv_image)
    cv2.waitKey(1)


def main(args=None):
    global bridge, node

    rclpy.init(args=args)

    print("Hello World, I am the cv image node![Python]\n")

    node = Node("cv_image_node")
    bridge = CvBridge()

    # create subscriber for RGB image
    node.create_subscription(
        Image,
        "/kinect2/qhd/image_raw",
        cam_rgb_callback,
        1
    )

    # creat a window
    cv2.namedWindow("RGB")
    cv2.namedWindow("test")

    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()



if __name__ == "__main__":
    main()

File: cv_pkg/src/cv_image_node.cpp
#include <rclcpp/rclcpp.hpp>
#include <cv_bridge/cv_bridge.h>
#include <sensor_msgs/msg/image.hpp>
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/highgui/highgui.hpp>

rclcpp::Node::SharedPtr node;


void Cam_RGB_Callback(const sensor_msgs::msg::Image::SharedPtr msg)
{
    cv_bridge::CvImagePtr cv_ptr;
    try
    {
        cv_ptr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::BGR8);
    }
    catch (cv_bridge::Exception& e)
    {
        RCLCPP_ERROR(node->get_logger(), "cv_bridge exception: %s", e.what());
        return;
    }

    cv::imshow("RGB", cv_ptr->image);
    cv::waitKey(1);
}

int main(int argc, char const *argv[])
{
    rclcpp::init(argc, argv);

    printf("Hello World, I am the cv image node!\n");

    node = rclcpp::Node::make_shared("cv_image_node");

    std::shared_ptr<rclcpp::Subscription<sensor_msgs::msg::Image>> rgb_sub =
        node->create_subscription<sensor_msgs::msg::Image>(
            "/kinect2/qhd/image_raw",
            1,
            Cam_RGB_Callback
        );

    cv::namedWindow("RGB");
    rclcpp::spin(node);
    cv::destroyAllWindows();
    rclcpp::shutdown();

    return 0;
}
